PXC-4012: Node leaves PXC cluster when running

kamil-holubicki opened the pull request:
concurrent CREATE USER with password_history option

https://jira.percona.com/browse/PXC-4012

Problem:
When password history is enabled and CREATE USER is executed in parallel, or one after another, but in the same second the replica node leaves the cluster.

Cause:
mysql.password_history has the primary key on User, Host, and Timestamp columns. Log_event_header carries the timestamp with one-second resolution. The information about the microseconds part of the timestamp is carried by Q_MICROSECONDS variable stored along with the event. However, as Q_MICROSECONDS is needed only for a few events, there is optimization related to THD::query_start_usec_used flag. This flag is set only for queries that require a high-precision timestamp when the query is processed. Then binlog event is logged and replicated (async replication).
With Galera replication in place, we replicate DDL as TOI. It happens at the very beginning of the query and THD::query_start_usec_used flag is not set yet, so the microseconds part of the timestamp is not sent.

Solution:
For user's password-related queries set THD::query_start_usec_used explicitly for the time of construction of Query_log_event which will be replicated. This is a temporary event, only for TOI replication.

kamil-holubicki committed:
PXC-4012: Node leaves PXC cluster when running
concurrent CREATE USER with password_history option

https://jira.percona.com/browse/PXC-4012

Problem:
When password history is enabled and CREATE USER is executed in
parallel, or one after another, but in the same second the replica node
leaves the cluster.

Cause:
mysql.password_history has the primary key on User, Host, and Timestamp
columns. Log_event_header carries the timestamp with one-second
resolution. The information about the microseconds part of the timestamp
is carried by Q_MICROSECONDS variable stored along with the event.
However, as Q_MICROSECONDS is needed only for a few events, there is
optimization related to THD::query_start_usec_used flag. This flag is
set only for queries that require a high-precision timestamp when the
query is processed. Then binlog event is logged and replicated (async
replication).
With Galera replication in place, we replicate DDL as TOI. It happens
at the very beginning of the query and THD::query_start_usec_used flag
is not set yet, so the microseconds part of the timestamp is not sent.

Solution:
For user's password-related queries set THD::query_start_usec_used
explicitly for the time of construction of Query_log_event which will be
replicated. This is a temporary event, only for TOI replication.

kamil-holubicki commented:
https://pxc.cd.percona.com/job/pxc-8.0-pipeline-parallel-mtr/681/

venkatesh-prasad-v commented on mysql-test/suite/galera/t/galera_password_history.test:
```diff
@@ -0,0 +1,46 @@
+# Test if replicated events which add rows to mysql.password_history table does not cause PK collision in this table.
+# The timestamp of the event is the part of PK. If replicated events miss usec part of the timestamp and they occur
+# on the same second, they will cause PK collision.
+# This test executes queries in bursts, so at list part of the burst should happen in the same second causing the collision.
```
```suggestion
# This test executes queries in bursts, so at least part of the burst should happen in the same second causing the collision.
```

kamil-holubicki replied:
ok

venkatesh-prasad-v commented on mysql-test/suite/galera/t/galera_password_history.test:
```diff
@@ -0,0 +1,46 @@
+SET PASSWORD FOR 'test1'@'%' = '*6C387FC38kzfld77969696969698698690888daf';
+SET PASSWORD FOR 'test1'@'%' = '*6C38fds8s8df8d88sf8d88sf8d8s8f8d8f8s8d8f';
+
+# check if mysql.password_history table was really used
+--let $assert_text = 'node_1: There should be 10 rows in mysql.password_history table.'
+--let $assert_cond = [SELECT COUNT(*) = 10 FROM mysql.password_history WHERE User = "test1"] = 1
+--source include/assert.inc
```
Can we validate that the contents are also same
```
--let $galera_diff_statement = SELECT * FROM mysql.password_history WHERE User = "test1";
--source include/galera_diff.inc
```

kamil-holubicki replied:
Yup, we can do this, but only for Host/User/Timestamp part. Password is kept as the hash and is unique to the server.

venkatesh-prasad-v commented on sql/wsrep_mysqld.cc:
```diff
@@ -1857,8 +1857,31 @@ int wsrep_to_buf_helper(THD *thd, const char *query, uint query_len,
+     the beginning of query processing.
+     Let's play safe, and change the flag only for the time of creation
+     of the event which is replicated. This way we will not affect non-wsrep
+     related logic. If it decides to cut precission, it will be still possible.
```
```suggestion
     related logic. If it decides to cut precision, it will be still possible.
```

kamil-holubicki replied:
ok

venkatesh-prasad-v commented on sql/wsrep_mysqld.cc:
```diff
@@ -1857,8 +1857,31 @@ int wsrep_to_buf_helper(THD *thd, const char *query, uint query_len,
   /* continue to append the actual query */
   Query_log_event ev(thd, query, query_len, false, false, false, 0);
+
+  /* Below commands will enter the row in mysql.password_history table.
```
Probably `insert` is more appropriate here?
```suggestion
  /* Below commands will insert the row in mysql.password_history table.
```

venkatesh-prasad-v commented on mysql-test/suite/galera/t/galera_password_history.cnf:
```diff
@@ -0,0 +1,4 @@
+!include ../galera_2nodes.cnf
+
+[mysqld]
+password_history=10
```
Not a strong opinion though! Can we move this part to user creation? This can reduce one file from our repo.
```
CREATE USER IF NOT EXISTS 'test1'@'%' IDENTIFIED WITH 'mysql_native_password' AS '*6C387FC3893DBA1E3BA155E74754DA6682D04747' PASSWORD HISTORY 10;
```

kamil-holubicki replied:
done

kamil-holubicki committed:
PXC-4012: Node leaves PXC cluster when running
concurrent CREATE USER with password_history option

https://jira.percona.com/browse/PXC-4012

Addressed review comments.
